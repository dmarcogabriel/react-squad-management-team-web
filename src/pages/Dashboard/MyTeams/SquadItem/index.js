import React from 'react';
import { MdShare, MdEdit } from 'react-icons/md';
import { FaTrashAlt } from 'react-icons/fa';
import cn from 'classnames';
import { Link } from 'react-router-dom';

const SquadItem = ({
  squad,
  onClick,
  onDelete,
  dataTestid,
  squadSelectButtonTestid,
  squadDeleteButtonTestid,
}) => (
  <tr>
    <td>
      <div
        data-testid={squadSelectButtonTestid || 'squadItem'}
        role="button"
        tabIndex={0}
        onClick={() => onClick(squad)}
        onKeyDown={(e) => e.keyCode === 13 && onClick(squad)}
        className={cn(
          'relative border-b-2 border-gray-200 rounded p-2 flex w-full',
          squad.selected && 'text-primary-dark bg-primary-light'
        )}
      >
        <p data-testid={dataTestid} className="inline mr-10">
          {squad.name}
        </p>

        <p className="inline ml-2">{squad.description.substring(0, 20)}</p>

        <div
          data-testid="actionButtons"
          className={cn(
            'flex items-center absolute h-full',
            !squad.selected && 'hidden',
            'right-0 top-0 mr-3'
          )}
        >
          <button
            data-testid={squadDeleteButtonTestid || 'squadItem-delete'}
            type="button"
            onClick={() => onDelete(squad.id)}
          >
            <FaTrashAlt />
          </button>

          <button type="button" className="mx-2">
            <MdShare />
          </button>

          <Link to={`create?squadId=${squad.id}`}>
            <MdEdit />
          </Link>
        </div>
      </div>
    </td>
  </tr>
);

export default SquadItem;
