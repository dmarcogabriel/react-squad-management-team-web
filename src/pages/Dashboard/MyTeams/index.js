import React, { useState, useEffect } from 'react';
import { MdAdd } from 'react-icons/md';
import { CgArrowsV } from 'react-icons/cg';
import { Link } from 'react-router-dom';
import cn from 'classnames';
import Card from '../../../common/Card';
import SquadItem from './SquadItem';
import { useSquad } from '../../../contexts/Squad';

const MyTeams = ({ className }) => {
  const [squads, setSquads] = useSquad();
  const [myTeams, setMyTeams] = useState([]);

  const handleSelect = (squad) => {
    setMyTeams(
      myTeams.map((team) => ({
        ...team,
        selected: team.id === squad.id,
      }))
    );
  };

  const handleDelete = (squadId) => {
    const newSquads = squads.filter((squad) => squad.id !== squadId);

    setMyTeams(newSquads);
    setSquads(newSquads);
  };

  const handleOrderList = (comparationKey) => {
    const sorted = myTeams.sort((pl1, pl2) =>
      pl1[comparationKey].localeCompare(pl2[comparationKey])
    );

    setMyTeams([...sorted]);
  };

  useEffect(() => {
    if (!myTeams.length) {
      setMyTeams(squads.map((squad) => ({ ...squad, selected: false })));
    }
  }, []);

  return (
    <Card
      dataTestid="container"
      className={className}
      title="My Teams"
      RightComponent={() => (
        <Link
          to="create"
          className={cn(
            'bg-gradient-to-t from-secondary-dark',
            ' to-secondary-light text-white p-1 rounded-lg'
          )}
        >
          <MdAdd size={22} />
        </Link>
      )}
    >
      <table className="w-full">
        <thead>
          <tr className="text-base flex">
            <th>
              <button
                data-testid="order-by-name"
                type="button"
                onClick={() => handleOrderList('name')}
                className={cn(
                  'flex items-center mx-3 py-2',
                  'border-r-2 border-gray-100'
                )}
              >
                <p className="font-normal mr-5">Name</p>

                <CgArrowsV className="mx-2" />
              </button>
            </th>
            <th>
              <button
                data-testid="order-by-description"
                type="button"
                onClick={() => handleOrderList('description')}
                className="flex items-center w-full mx-3 py-2"
              >
                <p className="font-normal mr-5">Description</p>

                <CgArrowsV className="mx-2" />
              </button>
            </th>
          </tr>
        </thead>

        <tbody data-testid="squad-list">
          {myTeams.map((squad, i) => (
            <SquadItem
              dataTestid={`squad-${i}`}
              squadSelectButtonTestid={`squad-select-${i}`}
              squadDeleteButtonTestid={`squad-delete-${i}`}
              key={squad.id}
              squad={squad}
              onClick={handleSelect}
              onDelete={handleDelete}
            />
          ))}
        </tbody>
      </table>
    </Card>
  );
};

export default MyTeams;
