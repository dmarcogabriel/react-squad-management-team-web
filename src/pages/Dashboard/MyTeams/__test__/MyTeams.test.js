import React from 'react';
import renderer from 'react-test-renderer';
import { fireEvent, render } from '@testing-library/react';
import ReactDOM from 'react-dom';
import { BrowserRouter as Router } from 'react-router-dom';
import MyTeams from '..';

jest.mock('contexts/Squad', () => {
  let squads = [
    {
      id: 's1',
      name: 'Test Squad D',
      description: 'This is a test B',
    },
    {
      id: 's2',
      name: 'Test Squad C',
      description: 'This is a test A',
    },
    {
      id: 's3',
      name: 'Test Squad A',
      description: 'This is a test D',
    },
    {
      id: 's4',
      name: 'Test Squad B',
      description: 'This is a test C',
    },
  ];

  const setSquads = (newSquads) => {
    squads = newSquads;
  };

  return {
    useSquad: () => [squads, setSquads],
  };
});

const Component = (props) => (
  <Router>
    <MyTeams {...props} />
  </Router>
);

describe('<MyTeams />', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');

    ReactDOM.render(<Component />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders classes correctly', () => {
    const { getByTestId } = render(<Component className="test-class" />);

    expect(getByTestId('container').className).toEqual(
      'bg-white rounded-lg shadow-lg my-10 w-full test-class'
    );
  });

  it('renders squad list correctly', () => {
    const { getByTestId } = render(<Component />);

    expect(getByTestId('squad-list').childElementCount).toEqual(4);
  });

  it('execute order by name function correctly', () => {
    const { getByTestId } = render(<Component />);

    fireEvent.click(getByTestId('order-by-name'));

    expect(getByTestId('squad-0')).toHaveTextContent('Test Squad A');
    expect(getByTestId('squad-3')).toHaveTextContent('Test Squad D');
  });

  it('execute order by description function correctly', () => {
    const { getByTestId } = render(<Component />);

    fireEvent.click(getByTestId('order-by-description'));

    expect(getByTestId('squad-0')).toHaveTextContent('Test Squad C');
    expect(getByTestId('squad-3')).toHaveTextContent('Test Squad A');
  });

  it('executes onClick function', () => {
    const { getByTestId } = render(<Component />);

    const selectButton = getByTestId('squad-select-0');
    fireEvent.click(selectButton);

    expect(selectButton.className).toEqual(
      'relative border-b-2 border-gray-200 rounded p-2 flex w-full text-primary-dark bg-primary-light'
    );
  });

  // todo: test onDelete function

  it('match snapshot', () => {
    expect(renderer.create(<Component />).toJSON()).toMatchSnapshot();
  });
});
