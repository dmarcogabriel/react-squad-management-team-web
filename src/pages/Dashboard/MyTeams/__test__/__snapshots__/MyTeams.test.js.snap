// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MyTeams /> match snapshot 1`] = `
<div
  className="bg-white rounded-lg shadow-lg my-10 w-full"
  data-testid="container"
>
  <div
    className="flex justify-between items-center p-8 border-b-2 border-gray-200"
  >
    <h1
      className="text-secondary-dark font-bold text-md"
      data-testid="card-title"
    >
      My Teams
    </h1>
    <a
      className="bg-gradient-to-t from-secondary-dark  to-secondary-light text-white p-1 rounded-lg"
      href="/create"
      onClick={[Function]}
    >
      <svg
        fill="currentColor"
        height={22}
        size={22}
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 24 24"
        width={22}
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
        />
      </svg>
    </a>
  </div>
  <div
    className="py-5 px-2"
    data-testid="card-children"
  >
    <table
      className="w-full"
    >
      <thead>
        <tr
          className="text-base flex"
        >
          <th>
            <button
              className="flex items-center mx-3 py-2 border-r-2 border-gray-100"
              data-testid="order-by-name"
              onClick={[Function]}
              type="button"
            >
              <p
                className="font-normal mr-5"
              >
                Name
              </p>
              <svg
                className="mx-2"
                fill="none"
                height="1em"
                stroke="currentColor"
                strokeWidth="0"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M7.75739 5.04077L9.1716 6.45498L11.0001 4.62652V10H13.0001V4.62661L14.8284 6.45498L16.2426 5.04077L12 0.798126L7.75739 5.04077Z"
                  fill="currentColor"
                />
                <path
                  d="M16.2426 18.9593L14.8284 17.545L13.0001 19.3734V14H11.0001V19.3735L9.1716 17.545L7.75739 18.9593L12 23.2019L16.2426 18.9593Z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </th>
          <th>
            <button
              className="flex items-center w-full mx-3 py-2"
              data-testid="order-by-description"
              onClick={[Function]}
              type="button"
            >
              <p
                className="font-normal mr-5"
              >
                Description
              </p>
              <svg
                className="mx-2"
                fill="none"
                height="1em"
                stroke="currentColor"
                strokeWidth="0"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M7.75739 5.04077L9.1716 6.45498L11.0001 4.62652V10H13.0001V4.62661L14.8284 6.45498L16.2426 5.04077L12 0.798126L7.75739 5.04077Z"
                  fill="currentColor"
                />
                <path
                  d="M16.2426 18.9593L14.8284 17.545L13.0001 19.3734V14H11.0001V19.3735L9.1716 17.545L7.75739 18.9593L12 23.2019L16.2426 18.9593Z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </th>
        </tr>
      </thead>
      <tbody
        data-testid="squad-list"
      />
    </table>
  </div>
</div>
`;
